lib/common/absprio.c

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2019 Fastly
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include <ctype.h>
+#include "h2o.h"
+#include "h2o/absprio.h"
+
+static void parse_urgency(const char *token, size_t token_len, uint8_t *urgency)
+{
+    /* Expects "u=[0=7]" */
+    if (token_len != 3 || token[1] != '=' || !('0' <= token[2] && token[2] <= '7'))
+        return;
+
+    *urgency = token[2] - '0';
+}
+
+static void parse_incremental(const char *token, size_t token_len, int *incremental)
+{
+    /* Expects "i=?(0|1)" */
+    if (token_len != 4 || token[1] != '=' || token[2] != '?' || !('0' == token[3] || '1' == token[3]))
+        return;
+
+    *incremental = token[3] - '0';
+}
+
+void h2o_absprio_parse_priority(const h2o_iovec_t *value, uint8_t *urgency, int *incremental)
+{
+    h2o_iovec_t iter = *value;
+    const char *token;
+    size_t token_len;
+
+    while ((token = h2o_next_token(&iter, ',', &token_len, NULL)) != NULL) {
```

Comment:
It is possible to obtain an `h2o_iovec_t` pointing to the value, if you pass an `h2o_iovec_t *` as the fourth argument.

```c
h2o_iovec_t value;
while ((token = h2o_next_token(&iter, ',', &token_len, &value)) != NULL) {
    if (token_len == 1 && token[0] == 'u') {
        if (value.base != NULL) {
            // value should contain '0' .. '7'
        } else {
            // no corresponding value, therefore it's an error
        }
    } else if (token_len == 1 && token[0] == 'i') {
        if (value.base != NULL) {
            // value should contain '?0' or '?1'
        } else {
            // value omitted, meaning that i is true
        }
    }
}
```

Reply:
👍 Was not aware of this feature. Let me incorporate your suggestion.

Also, I was totally missing out that sh-boolean value can be omitted if it's true. I'll include this in the test case as well.

Reply:
FWIW, I do not think you missed anything. Omission of "=?1" is a recent change to Structured Headers :-)